File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/model/Pagination.java
package com.taurushq.sdk.protect.client.model;

import java.util.regex.Pattern;

/**
 * Page-size constants, the one page-size resolver, and factories for request cursors.
 * <p>
 * Every list method follows the cross-SDK pagination contract: a page size is always
 * sent, {@code null} or {@code 0} means {@link #DEFAULT_PAGE_SIZE}, and a negative size
 * or one above {@link #MAX_PAGE_SIZE} is rejected with an {@link IllegalArgumentException}
 * before any request is made.
 * <p>
 * Offset lists return an {@link OffsetPagination}; cursor lists return a {@link CursorPage}
 * whose {@link CursorPage#getNextCursor()} is passed back until {@link CursorPage#hasMore()}
 * is false:
 * <pre>{@code
 * // Offset list
 * WalletResult wallets = client.getWalletService().getWallets(20, 0);
 * while (wallets.getPagination().hasMore()) {
 *     wallets = client.getWalletService().getWallets(20, wallets.getPagination().getNextOffset());
 * }
 *
 * // Cursor list
 * ChangeResult changes = client.getChangeService().getChanges(null, null, 20, null);
 * while (changes.getPage().hasMore()) {
 *     changes = client.getChangeService().getChanges(null, null, 20, changes.getPage().getNextCursor());
 * }
 * }</pre>
 */
public final class Pagination {

    /**
     * Page size sent when a caller leaves it unset or 0.
     */
    public static final int DEFAULT_PAGE_SIZE = 20;

    /**
     * Largest page size any list method accepts.
     */
    public static final int MAX_PAGE_SIZE = 100;

    /**
     * Largest price-history limit (one point a day for a year). That endpoint cannot page.
     */
    public static final int MAX_PRICE_HISTORY_LIMIT = 365;

    /**
     * Largest count the SDKs accept from the server: 2^53 - 1, the largest integer every
     * SDK (TypeScript included) represents exactly.
     */
    static final long MAX_COUNT = 9007199254740991L;

    /**
     * A canonical decimal count, as the server's uint64 fields arrive on the wire.
     */
    private static final Pattern COUNT = Pattern.compile("0|[1-9][0-9]{0,15}");

    /**
     * Longest server value echoed into an error message.
     */
    private static final int MAX_ECHO = 32;

    private Pagination() {
        // Static utility class
    }

    /**
     * Resolves a page size for a list method: {@code null} or 0 gives
     * {@link #DEFAULT_PAGE_SIZE}, a negative size or one above {@link #MAX_PAGE_SIZE} is
     * rejected.
     *
     * @param option the option name used in the error message
     * @param value  the caller's page size, may be null
     * @return the page size to send
     * @throws IllegalArgumentException if the size is negative or above the maximum
     */
    public static int resolvePageSize(final String option, final Integer value) {
        return resolveSize(option, value, MAX_PAGE_SIZE);
    }

    /**
     * The one page-size resolver: {@code null} or 0 gives {@link #DEFAULT_PAGE_SIZE}; a
     * negative value, or one above {@code maximum}, is rejected by name.
     *
     * @param option  the option name used in the error message
     * @param value   the caller's value, may be null
     * @param maximum the largest accepted value, or null when the endpoint has none
     * @return the value to send
     * @throws IllegalArgumentException if the value is negative or above the maximum
     */
    public static int resolveSize(final String option, final Integer value, final Integer maximum) {
        if (value == null || value == 0) {
            return DEFAULT_PAGE_SIZE;
        }
        if (maximum == null) {
            if (value < 0) {
                throw new IllegalArgumentException(option + " must not be negative, got " + value);
            }
        } else if (value < 0 || value > maximum) {
            throw new IllegalArgumentException(
                    option + " must be between 1 and " + maximum + ", got " + value);
        }
        return value;
    }

    /**
     * Resolves an offset for an offset list: it must not be negative.
     *
     * @param option the option name used in the error message
     * @param value  the caller's offset, 0 for the first page
     * @return the offset to send
     * @throws IllegalArgumentException if the offset is negative
     */
    public static long resolveOffset(final String option, final long value) {
        if (value < 0) {
            throw new IllegalArgumentException(option + " must not be negative, got " + value);
        }
        return value;
    }

    /**
     * Creates the request cursor for a page of a cursor list: the page size, plus the
     * previous page's {@link CursorPage#getNextCursor()} when there is one.
     * <p>
     * Without a cursor only the page size is sent (the first page); with one the SDK sends
     * {@code currentPage=<cursor>} and {@code pageRequest=NEXT}.
     *
     * @param pageSize the page size, null or 0 for {@link #DEFAULT_PAGE_SIZE}
     * @param cursor   a previous {@link CursorPage#getNextCursor()}, or null/empty for the
     *                 first page
     * @return the request cursor
     * @throws IllegalArgumentException if the page size is negative or above the maximum
     */
    public static ApiRequestCursor page(final Integer pageSize, final String cursor) {
        int size = resolvePageSize("pageSize", pageSize);
        if (cursor == null || cursor.isEmpty()) {
            return new ApiRequestCursor((PageRequest) null, size);
        }
        return new ApiRequestCursor(cursor, PageRequest.NEXT, size);
    }

    /**
     * Creates a cursor for the first page with default page size.
     *
     * @return a cursor for the first page
     */
    public static ApiRequestCursor first() {
        return first(DEFAULT_PAGE_SIZE);
    }

    /**
     * Creates a cursor for the first page with the specified page size.
     *
     * @param pageSize the number of items per page
     * @return a cursor for the first page
     * @throws IllegalArgumentException if pageSize is not between 1 and {@link #MAX_PAGE_SIZE}
     */
    public static ApiRequestCursor first(int pageSize) {
        return new ApiRequestCursor(PageRequest.FIRST, pageSize);
    }

    /**
     * Creates a cursor for the next page based on a response cursor.
     *
     * @param responseCursor the response cursor from the previous request
     * @param pageSize       the number of items per page
     * @return a cursor for the next page
     * @throws IllegalArgumentException if pageSize is invalid or responseCursor is null
     * @throws IllegalStateException    if there is no next page
     */
    public static ApiRequestCursor next(ApiResponseCursor responseCursor, int pageSize) {
        if (responseCursor == null) {
            throw new IllegalArgumentException("responseCursor cannot be null");
        }
        if (!responseCursor.hasNext()) {
            throw new IllegalStateException("No next page available");
        }
        return new ApiRequestCursor(responseCursor.getCurrentPage(), PageRequest.NEXT, pageSize);
    }

    /**
     * Creates a cursor for the next page with default page size.
     *
     * @param responseCursor the response cursor from the previous request
     * @return a cursor for the next page
     */
    public static ApiRequestCursor next(ApiResponseCursor responseCursor) {
        return next(responseCursor, DEFAULT_PAGE_SIZE);
    }

    /**
     * Creates a cursor for the previous page based on a response cursor.
     *
     * @param responseCursor the response cursor from the previous request
     * @param pageSize       the number of items per page
     * @return a cursor for the previous page
     * @throws IllegalArgumentException if pageSize is invalid or responseCursor is null
     * @throws IllegalStateException    if there is no previous page
     */
    public static ApiRequestCursor previous(ApiResponseCursor responseCursor, int pageSize) {
        if (responseCursor == null) {
            throw new IllegalArgumentException("responseCursor cannot be null");
        }
        if (!responseCursor.hasPrevious()) {
            throw new IllegalStateException("No previous page available");
        }
        return new ApiRequestCursor(responseCursor.getCurrentPage(), PageRequest.PREVIOUS, pageSize);
    }

    /**
     * Creates a cursor for the previous page with default page size.
     *
     * @param responseCursor the response cursor from the previous request
     * @return a cursor for the previous page
     */
    public static ApiRequestCursor previous(ApiResponseCursor responseCursor) {
        return previous(responseCursor, DEFAULT_PAGE_SIZE);
    }

    /**
     * Creates a cursor for the last page with the specified page size.
     *
     * @param pageSize the number of items per page
     * @return a cursor for the last page
     * @throws IllegalArgumentException if pageSize is not between 1 and {@link #MAX_PAGE_SIZE}
     */
    public static ApiRequestCursor last(int pageSize) {
        return new ApiRequestCursor(PageRequest.LAST, pageSize);
    }

    /**
     * Creates a cursor for the last page with default page size.
     *
     * @return a cursor for the last page
     */
    public static ApiRequestCursor last() {
        return last(DEFAULT_PAGE_SIZE);
    }

    /**
     * Creates a cursor with a specific page token for direct navigation.
     *
     * @param pageToken   the page token to navigate to
     * @param pageRequest the type of page request
     * @param pageSize    the number of items per page
     * @return a cursor for the specified page
     * @throws IllegalArgumentException if pageSize is not between 1 and {@link #MAX_PAGE_SIZE}
     */
    public static ApiRequestCursor of(String pageToken, PageRequest pageRequest, int pageSize) {
        return new ApiRequestCursor(pageToken, pageRequest, pageSize);
    }

    /**
     * Checks an explicit page size, as carried by an {@link ApiRequestCursor}.
     *
     * @param pageSize the page size
     * @return the same page size
     * @throws IllegalArgumentException if it is not between 1 and {@link #MAX_PAGE_SIZE}
     */
    static long checkPageSize(final long pageSize) {
        if (pageSize < 1 || pageSize > MAX_PAGE_SIZE) {
            throw new IllegalArgumentException(
                    "pageSize must be between 1 and " + MAX_PAGE_SIZE + ", got " + pageSize);
        }
        return pageSize;
    }

    /**
     * Parses a count from a reply. Absent means 0, because the server omits zero values;
     * anything but a canonical decimal in [0, 2^53 - 1] is a malformed reply.
     *
     * @param field the reply field, for the error message
     * @param value the value as received, may be null
     * @return the count
     * @throws IntegrityException if the value is not a canonical count in range
     */
    static long parseCount(final String field, final String value) {
        if (value == null) {
            return 0L;
        }
        if (COUNT.matcher(value).matches()) {
            long parsed = Long.parseLong(value);
            if (parsed <= MAX_COUNT) {
                return parsed;
            }
        }
        String shown = value.length() > MAX_ECHO ? value.substring(0, MAX_ECHO) + "..." : value;
        throw new IntegrityException("malformed " + field + " in the server reply: \"" + shown
                + "\" is not a count between 0 and " + MAX_COUNT);
    }
}

File: taurus-protect-sdk-go/pkg/protect/service/pagination.go
package service

import (
	"fmt"
	"math"
	"strconv"

	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// The pagination contract shared by every list method in all four SDKs (repo-root CLAUDE.md,
// "Pagination (cross-SDK)"). scripts/resources/pagination-vectors.json pins the helpers below and
// list-request-vectors.json pins what each list method sends.
//
//	options ─▶ resolve*Window (page size 0→20, >100 or <0 → error) ─▶ request
//	reply ────▶ offsetPagination(rule) / cursorPage ─▶ model value
//
// validatord omits zero values, so an absent totalItems/offset/hasNext/next means 0/false/none and
// `{}` is a valid empty page.

// maxCount is the largest count a reply may carry. 2^53-1 keeps all four SDKs, TypeScript's
// number included, reading the same value.
const maxCount = 1<<53 - 1

// pageRequestNext is the pageRequest a continuation sends. validatord matches it exactly.
const pageRequestNext = "NEXT"

// resolvePageSize applies the shared page-size rule to a list option.
func resolvePageSize(option string, size int64) (int64, error) {
	return resolveSize(option, size, model.MaxPageSize)
}

// resolveSize maps 0 to model.DefaultPageSize and rejects a negative size or one above max.
// max 0 means the endpoint has no SDK maximum.
func resolveSize(option string, size, max int64) (int64, error) {
	if size < 0 || (max > 0 && size > max) {
		if max > 0 {
			return 0, fmt.Errorf("invalid %s %d: must be between 1 and %d (0 selects the default %d)",
				option, size, max, model.DefaultPageSize)
		}
		return 0, fmt.Errorf("invalid %s %d: must not be negative (0 selects the default %d)",
			option, size, model.DefaultPageSize)
	}
	if size == 0 {
		return model.DefaultPageSize, nil
	}
	return size, nil
}

// offsetWindow is the limit/offset pair an offset list sends.
type offsetWindow struct {
	limit  int64
	offset int64
}

// resolveOffsetWindow validates Limit/Offset before any request is sent.
func resolveOffsetWindow(limit, offset int64) (offsetWindow, error) {
	size, err := resolvePageSize("Limit", limit)
	if err != nil {
		return offsetWindow{}, err
	}
	if offset < 0 {
		return offsetWindow{}, fmt.Errorf("invalid Offset %d: must not be negative", offset)
	}
	return offsetWindow{limit: size, offset: offset}, nil
}

// offsetQueryRequest is satisfied by every generated request builder taking limit/offset.
type offsetQueryRequest[T any] interface {
	Limit(string) T
	Offset(string) T
}

// applyOffsetWindow always sends the limit, and the offset when it is not 0.
func applyOffsetWindow[T offsetQueryRequest[T]](req T, w offsetWindow) T {
	req = req.Limit(strconv.FormatInt(w.limit, 10))
	if w.offset > 0 {
		req = req.Offset(strconv.FormatInt(w.offset, 10))
	}
	return req
}

// offsetRule is how an endpoint's next page offset is derived; each offset endpoint has one.
type offsetRule int

const (
	// ruleReplyOffset: the reply's offset IS the next page's offset (offset + rows when absent).
	// WalletService_GetWalletsV2, WalletService_GetAddresses.
	ruleReplyOffset offsetRule = iota + 1
	// rulePlusRows: offset + rows. Transactions, fee payers, actions.
	rulePlusRows
	// rulePlusMinRowsLimit: offset + min(rows, limit); a synthetic daemon user or technical
	// group can be appended beyond the limit. Users, groups.
	rulePlusMinRowsLimit
	// rulePlusServerRows: offset + the rows the SERVER returned, before SDK exclusions; rows the
	// server dropped for a bad signature are gone for good. Whitelisted addresses.
	rulePlusServerRows
	// rulePlusLimit: offset + limit; rows the server skipped keep their SQL slot, so a short page
	// is not the end. Whitelisted contracts (assets).
	rulePlusLimit
)

// offsetReply is what an offset reply says about paging.
type offsetReply struct {
	TotalItems *string
	Offset     *string
}

// offsetPagination builds the page value every offset list returns.
//
// servedRows is the number of rows the server returned; excluded the rows the SDK withheld from
// them. TotalItems is reduced by the exclusions, NextOffset and HasMore never are: pagination
// position is a server-side fact. HasMore also requires progress, so a trailing empty page under
// an upper-bound total ends a walk.
func offsetPagination(rule offsetRule, w offsetWindow, servedRows, excluded int, reply offsetReply) (*model.Pagination, error) {
	total, err := parseCount("totalItems", reply.TotalItems)
	if err != nil {
		return nil, err
	}

	rows := int64(servedRows)
	var next int64
	switch rule {
	case ruleReplyOffset:
		if reply.Offset == nil {
			next = addSaturated(w.offset, rows)
		} else if next, err = parseCount("offset", reply.Offset); err != nil {
			return nil, err
		}
	case rulePlusRows, rulePlusServerRows:
		next = addSaturated(w.offset, rows)
	case rulePlusMinRowsLimit:
		next = addSaturated(w.offset, min(rows, w.limit))
	case rulePlusLimit:
		next = addSaturated(w.offset, w.limit)
	default:
		return nil, fmt.Errorf("unknown offset pagination rule %d", rule)
	}

	reported := total - int64(excluded)
	if reported < 0 {
		reported = 0
	}
	return &model.Pagination{
		Limit:      w.limit,
		Offset:     w.offset,
		TotalItems: reported,
		NextOffset: next,
		HasMore:    next > w.offset && next < total,
	}, nil
}

// cursorWindow is what a cursor list sends: always a page size, plus either a continuation
// (the caller's Cursor as currentPage with pageRequest NEXT) or the low-level
// CurrentPage/PageRequest options as given.
type cursorWindow struct {
	pageSize    int64
	currentPage string
	pageRequest string
}

// resolveCursorWindow validates the paging options of a cursor list. Cursor always means "the
// page after this one", so combining it with a low-level option is rejected rather than letting
// one of them silently lose.
func resolveCursorWindow(pageSize int64, cursor, currentPage, pageRequest string) (cursorWindow, error) {
	size, err := resolvePageSize("PageSize", pageSize)
	if err != nil {
		return cursorWindow{}, err
	}
	if cursor == "" {
		return cursorWindow{pageSize: size, currentPage: currentPage, pageRequest: pageRequest}, nil
	}
	if currentPage != "" {
		return cursorWindow{}, fmt.Errorf("invalid options: Cursor and CurrentPage cannot both be set; Cursor continues from a previous page's NextCursor")
	}
	if pageRequest != "" {
		return cursorWindow{}, fmt.Errorf("invalid options: Cursor and PageRequest cannot both be set; Cursor always requests the %s page", pageRequestNext)
	}
	return cursorWindow{pageSize: size, currentPage: cursor, pageRequest: pageRequestNext}, nil
}

// cursorQueryRequest is satisfied by every generated request builder taking cursor.* parameters.
type cursorQueryRequest[T any] interface {
	CursorCurrentPage(string) T
	CursorPageRequest(string) T
	CursorPageSize(string) T
}

// applyCursorQuery sends the window as cursor.* query parameters.
func applyCursorQuery[T cursorQueryRequest[T]](req T, w cursorWindow) T {
	req = req.CursorPageSize(strconv.FormatInt(w.pageSize, 10))
	if w.currentPage != "" {
		req = req.CursorCurrentPage(w.currentPage)
	}
	if w.pageRequest != "" {
		req = req.CursorPageRequest(w.pageRequest)
	}
	return req
}

// requestCursorQueryRequest is satisfied by generated builders taking requestCursor.* parameters.
type requestCursorQueryRequest[T any] interface {
	RequestCursorCurrentPage(string) T
	RequestCursorPageRequest(string) T
	RequestCursorPageSize(string) T
}

// applyRequestCursorQuery sends the window as requestCursor.* query parameters.
func applyRequestCursorQuery[T requestCursorQueryRequest[T]](req T, w cursorWindow) T {
	req = req.RequestCursorPageSize(strconv.FormatInt(w.pageSize, 10))
	if w.currentPage != "" {
		req = req.RequestCursorCurrentPage(w.currentPage)
	}
	if w.pageRequest != "" {
		req = req.RequestCursorPageRequest(w.pageRequest)
	}
	return req
}

// body renders the window for endpoints that take the cursor in the request body.
func (w cursorWindow) body() *openapi.TgvalidatordRequestCursor {
	size := strconv.FormatInt(w.pageSize, 10)
	c := &openapi.TgvalidatordRequestCursor{PageSize: &size}
	if w.currentPage != "" {
		currentPage := w.currentPage
		c.CurrentPage = &currentPage
	}
	if w.pageRequest != "" {
		pageRequest := w.pageRequest
		c.PageRequest = &pageRequest
	}
	return c
}

// cursorReply is what a cursor-paged reply says about paging.
type cursorReply struct {
	// Cursor is the reply cursor of the cursor.* and requestCursor lists.
	Cursor *openapi.TgvalidatordResponseCursor
	// TokenOnly marks the lists that page by a bare token instead (wallet tokens, rules
	// history); Token is the reply's next-page token.
	TokenOnly bool
	Token     *string
	// HasTotal marks the lists whose reply carries a count; an absent Total is then 0.
	HasTotal bool
	Total    *string
	// Excluded is the number of rows the SDK withheld; it reduces the total (never below 0).
	Excluded int
}

// cursorPage builds the page value every cursor list returns.
//
// A cursor list continues with the reply's currentPage only when hasNext is set: sending NEXT
// past the end is a 400 on the v2 lists and silently returns page 1 on the keyset lists, so a
// cursor is never derived any other way. A token-only list continues with its token, present
// exactly when another page exists.
func cursorPage(pageSize int64, reply cursorReply) (model.CursorPage, error) {
	page := model.CursorPage{PageSize: pageSize}
	if reply.HasTotal {
		total, err := parseCount("total", reply.Total)
		if err != nil {
			return model.CursorPage{}, err
		}
		total = max(total-int64(reply.Excluded), 0)
		page.TotalItems = &total
	}

	if reply.TokenOnly {
		if reply.Token != nil && *reply.Token != "" {
			page.HasMore = true
			page.NextCursor = *reply.Token
		}
		return page, nil
	}

	c := reply.Cursor
	if c == nil || c.HasNext == nil || !*c.HasNext {
		return page, nil
	}
	if c.CurrentPage == nil || *c.CurrentPage == "" {
		return model.CursorPage{}, fmt.Errorf("%w: the reply cursor has hasNext but no currentPage", model.ErrMalformedPagination)
	}
	page.HasMore = true
	page.NextCursor = *c.CurrentPage
	return page, nil
}

// parseCount reads a count from a reply. Absent means 0; anything but a canonical decimal in
// [0, 2^53-1] is an error, because a count silently read as 0 ends a walk early.
func parseCount(field string, value *string) (int64, error) {
	if value == nil {
		return 0, nil
	}
	s := *value
	if !isCanonicalDecimal(s) {
		return 0, fmt.Errorf("%w: reply %s %q is not a decimal count", model.ErrMalformedPagination, field, s)
	}
	n, err := strconv.ParseUint(s, 10, 64)
	if err != nil || n > maxCount {
		return 0, fmt.Errorf("%w: reply %s %q is above %d", model.ErrMalformedPagination, field, s, uint64(maxCount))
	}
	return int64(n), nil
}

// isCanonicalDecimal accepts digits only, without a sign or a leading zero.
func isCanonicalDecimal(s string) bool {
	if s == "" || (len(s) > 1 && s[0] == '0') {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// addSaturated adds a non-negative b to a, stopping at math.MaxInt64.
func addSaturated(a, b int64) int64 {
	if b > 0 && a > math.MaxInt64-b {
		return math.MaxInt64
	}
	return a + b
}

// chunkIDs splits ids into consecutive batches of at most size, so an id-filtered internal read
// stays within both the page-size maximum and the endpoint's own id cap.
func chunkIDs(ids []string, size int) [][]string {
	var chunks [][]string
	for start := 0; start < len(ids); start += size {
		end := min(start+size, len(ids))
		chunks = append(chunks, ids[start:end])
	}
	return chunks
}
